### ISSUE TITLE
Function calls with side effects are removed (when they shouldn't be)

### ISSUE THREAD

**Issue by Pauan** [rollup/rollup#476]
Here is a reduced test case:

http://rollupjs.org/#%7B%22options%22%3A%7B%22format%22%3A%22amd%22%2C%22moduleName%22%3A%22myBundle%22%2C%22globals%22%3A%7B%7D%2C%22moduleId%22%3A%22%22%7D%2C%22modules%22%3A%5B%7B%22name%22%3A%22main.js%22%2C%22code%22%3A%22const%20foo%20%3D%20%7B%7D%3B%5Cn%5Cnconst%20set%20%3D%20(key%2C%20value)%20%3D%3E%20%7B%5Cn%20%20foo%5Bkey%5D%20%3D%20value%3B%5Cn%7D%3B%5Cn%5Cnset(%5C%22foo%5C%22%2C%201)%3B%5Cnset(%5C%22bar%5C%22%2C%202)%3B%5Cnset(%5C%22qux%5C%22%2C%203)%3B%5Cn%5Cnconsole.log(foo)%3B%22%7D%5D%7D

As you can see, it completely removes the calls to `set`

In my real application, I am using an object to store configuration, and then I am using function calls to populate the object.

But rollup removes all of the function calls, so the configuration ends up being empty.

**Comment by Victorystick** [comment]
@Pauan Thank you for reporting this issue.

For reference, here is the offending code:
```js
const foo = {};

const set = (key, value) => {
  foo[key] = value;
};

set("foo", 1);
set("bar", 2);
set("qux", 3);

console.log(foo);
```

It's curious that the resulting core preserves the _definition_ of `set` as it's never used in the output. Removing the `console.log` also results in the definition of `set` being removed.

Replacing the `set` arrow-function _expression_ with a function _declaration_ like so, `set` is [completely removed from the output](http://rollupjs.org/#%7B%22options%22%3A%7B%22format%22%3A%22amd%22%2C%22moduleName%22%3A%22myBundle%22%2C%22globals%22%3A%7B%7D%2C%22moduleId%22%3A%22%22%7D%2C%22modules%22%3A%5B%7B%22name%22%3A%22main.js%22%2C%22code%22%3A%22const%20foo%20%3D%20%7B%7D%3B%5Cn%5Cnfunction%20set(%20key%2C%20value%20)%20%7B%5Cn%20%20foo%5Bkey%5D%20%3D%20value%3B%5Cn%7D%3B%5Cn%5Cnset(%5C%22foo%5C%22%2C%201)%3B%5Cnset(%5C%22bar%5C%22%2C%202)%3B%5Cnset(%5C%22qux%5C%22%2C%203)%3B%5Cn%5Cnconsole.log(foo)%3B%22%7D%5D%7D).
```js
const foo = {};

function set( key, value ) {
  foo[key] = value;
}

set("foo", 1);
set("bar", 2);
set("qux", 3);

console.log(foo);
```

I suspect that Rollup preserves the definition of the arrow function above, as it believes that the _definition itself_ somehow _modifies_ `foo`; while the calls to `set` don't. I've no idea why every reference to `set` is removed in the second example.

**Comment by eventualbuddha** [comment]
Interestingly, commenting out `const foo = {};` also "fixes" it. I tried to go back in the rollup git history to see when this broke but I can't always get it to reliably build. Perhaps @Rich-Harris can explain why I'm getting missing modules (i.e. minimist, ansi-styles) with this script:

```sh
#!/usr/bin/env bash

set -e

cat <<EOS > file.js
const foo = {};

const set = (key, value) => {
  foo[key] = value;
};

set("foo", 1);
set("bar", 2);
set("qux", 3);

console.log(foo);
EOS

git clean -dfx test
npm install
npm run build
./bin/rollup file.js | grep qux
echo PASS
```

**Comment by kzc** [comment]
Rollup is not acknowledging the side effect in the MemberExpression of the left side of the AssignmentExpression in function `set`, which is why the calls to `set` are thought to be side-effect-free and excluded.

If the test case is modified as follows it works as one would hope:

```js
const foo = {};
function set(obj, key, value) {
  obj[key] = value;
}
set(foo,"bar", 2);
set(foo,"qux", 3);
```

The test above was was constructed after observing that a side effect is generated from a modifier node that happens to be a parameter:

https://github.com/rollup/rollup/blob/4b6145574de33d3af5957f37d7eefd8afbdc91b3/src/utils/run.js#L97

I think a fix has to be introduced in the general vicinity of this code to get rollup to recognize the side effect in the foo AssignmentExpression MemberExpression which is in the same module but outside the block scope of the function.

**Comment by kzc** [comment]
The same reason why rollup presently drops `obj` and `obj.foo` from this test case:

    var obj = {};
    obj.foo = function() { console.log('obj.foo() called'); }
    function bar() { console.log('bar() called'); }
    bar();

to produce:

    function bar() { console.log('bar() called'); }
    bar();

is also why rollup does not acknowledge a side effect in the `set` function in:

    const foo = {};
    function set(key, value) {
      foo[key] = value;
    }
    set(foo, "bar", 2);
    set(foo, "qux", 3);
    console.log(foo);

which ultimately causes the `set()` calls to be incorrectly dropped. 

In both cases the condition below marked as `(*)` was false in `run()`, and the side effect not triggered:

			else if ( isModifierNode( node ) ) {
				let subject = node[ modifierNodes[ node.type ] ];
				while ( subject.type === 'MemberExpression' ) subject = subject.object;

				let declaration = scope.findDeclaration( subject.name );

				if ( declaration ) {
					if ( declaration.isParam ) hasSideEffect = true;
				} else {
					declaration = statement.module.trace( subject.name );

					if ( !declaration || declaration.isExternal || declaration.isUsed ) {
						// (*) Not reached for member expression assignments
						//     for both test cases.
						hasSideEffect = true;
					}
				}
			}

https://github.com/rollup/rollup/blob/4b6145574de33d3af5957f37d7eefd8afbdc91b3/src/utils/run.js#L90-L102

Some code has to crafted to differentiate whether the declaration in question is actually used by the final program.

**Comment by kzc** [comment]
Fix in PR #512

**Comment by Pauan** [comment]
Awesome, thanks guys!
